### ISSUE TITLE
FastTree has a reference to non-existing assembly

### ISSUE THREAD

**Issue by eerhardt** [dotnet/machinelearning#1443]
We currently have a bunch of P/Invokes to an assembly that can't be found:

https://github.com/dotnet/machinelearning/blob/69a549e5119a63c1740cdf2526098337b42a76f5/src/Microsoft.ML.FastTree/BinFile/IniFileParserInterface.cs#L13-L20

When I try running the following code:

```C#
        static void Main(string[] args)
        {
            var env = new LocalEnvironment();

            var data = new TextLoader(env, new TextLoader.Arguments
            {
                HasHeader = true,
                Separator = "\t",
                Column = new[]
                {
                    new TextLoader.Column("Label", DataKind.R4, 0),
                    new TextLoader.Column("Workclass", DataKind.Text, 1),
                    new TextLoader.Column("NumericFeatures", DataKind.R4, new [] { new TextLoader.Range(9, 14) })
                }
            }).Read(new MultiFileSource(@"F:\git\machinelearning\test\data\adult.tiny.with-schema.txt"));

            // Pipeline.
            var pipeline = new TermEstimator(env, new[]{
                                    new TermTransform.ColumnInfo("Workclass", "Group"),
                                    new TermTransform.ColumnInfo("Label", "Label0") })
                .Append(new FastTreeRankingTrainer(env, "Label0", "NumericFeatures", "Group",
                    advancedSettings: s => { s.NumTrees = 10; s.PositionDiscountFreeform = "foo"; }));

            pipeline.Fit(data);
        }
```

I get the exception:

```

Unhandled Exception: System.DllNotFoundException: Unable to load DLL 'NeuralTreeEvaluator.dll': The specified module could not be found. (Exception from HRESULT: 0x8007007E)
   at Microsoft.ML.Runtime.FastTree.Internal.IniFileParserInterface.Native.CreateFromFreeform(String freeform)
   at Microsoft.ML.Runtime.FastTree.Internal.IniFileParserInterface.CreateFromFreeform(String freeform)
   at Microsoft.ML.Runtime.FastTree.FastTreeRankingTrainer.LambdaRankObjectiveFunction.FillDiscounts(String positionDiscountFreeform)
   at Microsoft.ML.Runtime.FastTree.FastTreeRankingTrainer.LambdaRankObjectiveFunction..ctor(Dataset trainset, Int16[] labels, Arguments args, IParallelTraining parallelTraining)
   at Microsoft.ML.Runtime.FastTree.FastTreeRankingTrainer.ConstructObjFunc(IChannel ch)
   at Microsoft.ML.Runtime.FastTree.BoostingFastTreeTrainerBase`3.ConstructOptimizationAlgorithm(IChannel ch)
   at Microsoft.ML.Runtime.FastTree.FastTreeRankingTrainer.ConstructOptimizationAlgorithm(IChannel ch)
   at Microsoft.ML.Runtime.FastTree.FastTreeTrainerBase`3.Initialize(IChannel ch)
   at Microsoft.ML.Runtime.FastTree.FastTreeRankingTrainer.Initialize(IChannel ch)
   at Microsoft.ML.Runtime.FastTree.FastTreeTrainerBase`3.TrainCore(IChannel ch)
   at Microsoft.ML.Runtime.FastTree.FastTreeRankingTrainer.TrainModelCore(TrainContext context)
   at Microsoft.ML.Runtime.Training.TrainerEstimatorBase`2.TrainTransformer(IDataView trainSet, IDataView validationSet, IPredictor initPredictor)
   at Microsoft.ML.Runtime.Data.EstimatorChain`1.Fit(IDataView input)
   at ConsoleApp46.Program.Main(String[] args) in C:\Users\eerhardt\source\repos\ConsoleApp46\ConsoleApp46\Program.cs:line 38
```

/cc @codemzs

**Comment by codemzs** [comment]
yikes we need to remove this bin loader that is used for a specific file format only used by Bing.,

**Comment by mstfbl** [comment]
Hi @eerhardt , I have updated and rewrote the code sample you've provided as a unit test in `TermEstimatorTests.cs` as such:

>         [Fact]
>         public void TempTestIssue1443()
>         {
>             string dataPath = GetDataPath("adult.tiny.with-schema.txt");
> 
>             var loader = new TextLoader(ML, new TextLoader.Options
>             {
>                 Columns = new[]{
>                     new TextLoader.Column("Label", DataKind.Single, 0),
>                     new TextLoader.Column("Workclass", DataKind.String, 1),
>                     new TextLoader.Column("NumericFeatures", DataKind.Single, new [] { new TextLoader.Range(9, 14) })
>                 },
>                 Separator = "\t",
>                 HasHeader = true
>             }, new MultiFileSource(dataPath));
> 
>             var est = new ValueToKeyMappingEstimator(Env, new[]{
>                     new ValueToKeyMappingEstimator.ColumnOptions("Group", "Workclass"),
>                     new ValueToKeyMappingEstimator.ColumnOptions("Label0", "Label")})
>                 .Append(new FastTreeRankingTrainer(ML, "Label0", "NumericFeatures", "Group", null, 10));
>             var data = loader.Load(dataPath);
>             var transformer = est.Fit(data);
>         }

This test ran successfully, and I did not encounter the error you provided, nor did the call stack go anywhere in `IniFileParserInterface.cs`. Am I missing something here? Perhaps `IniFileParserInterface` is no longer used?

**Comment by eerhardt** [comment]
Your test code isn’t the same as the code as I posted above.

Look at the CreateFromRanker code I linked above and find all references to see how it is used.

**Comment by mstfbl** [comment]
I was able to reproduce the error with a static call to `IniFileParserInterface.CreateFromRanker(string path)`, though I don't understand where the code I provided differs.

I agree with Zeeshan that `src/Microsoft.ML.FastTree/BinFile/` needs to be removed, and I'll be doing this in an upcoming PR.

**Comment by eerhardt** [comment]
The code at the top sets ‘ PositionDiscountFreeform’, which is what was triggering the dllimport. It looks like this option may be internal in the latest code. (This issue is 18 months old)

**Comment by mstfbl** [comment]
Fixed with PR #5046 .
